### PR TITLE
[RL] support update_weights_from_tensor for mtp

### PR DESCRIPTION
**Pull request by zhuzilin**
## Motivation

This PR does 2 things:

~1. Put the `monkey_patch_torch_reductions` into `TpModelWorker.update_weights_from_tensor` to greatly accelerate the update.~
2. Add `update_weights_from_tensor` to `EAGLEWorker` so we can update the mtp head of a model.

Thank you your your time on reviewing this PR :)

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Review by zhaochenyang20 — CHANGES_REQUESTED** [review]
unit test for it?

File: python/sglang/srt/managers/tp_worker.py [hunk 1]
```diff
@@ -262,6 +263,7 @@ def update_weights_from_distributed(
         return success, message
 
     def update_weights_from_tensor(self, recv_req: UpdateWeightsFromTensorReqInput):
+        monkey_patch_torch_reductions()
```
Comment (hebiao064) [inline python/sglang/srt/managers/tp_worker.py]: iiuc https://github.com/pytorch/pytorch/pull/149248 this monkey patch function is to solve the wrong cuda device problem, I wonder why it could have impact on performance?

File: python/sglang/srt/speculative/eagle_worker.py [hunk 2]
```diff
@@ -15,6 +15,10 @@
 )
 from sglang.srt.layers.logits_processor import LogitsProcessorOutput
 from sglang.srt.layers.sampler import get_token_ids_logprobs, get_top_logprobs
+from sglang.srt.managers.io_struct import (
+    UpdateWeightsFromDistributedReqInput,
```
Comment (hebiao064) [inline python/sglang/srt/speculative/eagle_worker.py]: UpdateWeightsFromDistributedReqInput is this being used?

File: python/sglang/srt/managers/tp_worker.py (hunk 1 above)
Comment (zhuzilin) [reply python/sglang/srt/managers/tp_worker.py]: hmm... I think it related to the ray usage of the rl framework, as ray will rewrite the `CUDA_VISIBLE_DEVICES` of each process, and without this patch, there may be unecessary copy during weight updation, because of the following function in model_runner.py:

```bash
def _unwrap_tensor(tensor, tp_rank):
    if isinstance(tensor, LocalSerializedTensor):
        monkey_patch_torch_reductions()
        tensor = tensor.get(tp_rank)
    return tensor.to(torch.cuda.current_device())
```

File: python/sglang/srt/speculative/eagle_worker.py (hunk 2 above)
Comment (zhuzilin) [reply python/sglang/srt/speculative/eagle_worker.py]: removed.

File: python/sglang/srt/managers/tp_worker.py (hunk 1 above)
Comment (zhuzilin) [reply python/sglang/srt/managers/tp_worker.py]: I found that as long as the rl framework sets `RAY_EXPERIMENTAL_NOSET_CUDA_VISIBLE_DEVICES` to prevent ray from fiddling with `CUDA_VISIBLE_DEVICES`, we don't need to add the monkey patch.

**Review by hebiao064 — APPROVED** [review]

File: python/sglang/srt/managers/scheduler.py [hunk 3]
```diff
@@ -2326,7 +2326,10 @@ def update_weights_from_distributed(
 
     def update_weights_from_tensor(self, recv_req: UpdateWeightsFromTensorReqInput):
         """Update the online model parameter from tensors."""
-        success, message = self.tp_worker.update_weights_from_tensor(recv_req)
+        if self.draft_worker is not None:
+            success, message = self.draft_worker.update_weights_from_tensor(recv_req)
+        else:
+            success, message = self.tp_worker.update_weights_from_tensor(recv_req)
```
Comment (hebiao064) [inline python/sglang/srt/managers/scheduler.py]: when spec decoding is enabled, how to update weights from tensor for the main tp worker?

Comment (fzyzcjy) [reply python/sglang/srt/managers/scheduler.py]: +1 it seems we need to handle that as well

**Review by fzyzcjy — CHANGES_REQUESTED** [review]
general idea LGTM, just one nit

**Comment by fzyzcjy** [comment]
after the nit above (and resolving conflicts) I think it is ready to merge

**Comment by guapisolo** [comment]
> after the nit above (and resolving conflicts) I think it is ready to merge

@fzyzcjy A small problem.  

The RL framework may not support update weight for external draft model. e.g. use lmsys/sglang-EAGLE3-LLaMA3.1-Instruct-8B to speed up inference speed only (without training draft model). 

So I think `if self.draft_worker is not None:` is not a good condition, may be we should add another parameter to control whether update draft model weights or not.

File: python/sglang/srt/managers/scheduler_update_weights_mixin.py [hunk 4]
```diff
@@ -76,7 +76,10 @@ def update_weights_from_distributed(
 
     def update_weights_from_tensor(self, recv_req: UpdateWeightsFromTensorReqInput):
         """Update the online model parameter from tensors."""
-        success, message = self.tp_worker.update_weights_from_tensor(recv_req)
+        if self.draft_worker is not None:
+            success, message = self.draft_worker.update_weights_from_tensor(recv_req)
+        else:
+            success, message = self.tp_worker.update_weights_from_tensor(recv_req)
```
Comment (fzyzcjy) [inline python/sglang/srt/managers/scheduler_update_weights_mixin.py]:
```suggestion
        worker = self.draft_worker or self.tp_worker
        success, message = worker.update_weights_from_tensor(recv_req)
```

**Review by fzyzcjy — APPROVED** [review]

File: python/sglang/srt/managers/scheduler_update_weights_mixin.py (hunk 4 above)
Comment (zhuzilin) [reply python/sglang/srt/managers/scheduler_update_weights_mixin.py]: fixed.
